src/Lykke.Job.CandlesHistoryWriter.Services/Candles/RedisCacheCaretaker.cs

```diff
@@ -0,0 +1,82 @@
+using StackExchange.Redis;
+
+namespace Lykke.Job.CandlesHistoryWriter.Services.Candles
+{
+    [UsedImplicitly]
+    public class RedisCacheCaretaker : TimerPeriod
+    {
+        private readonly ICandlesHistoryRepository _historyRepository;
+        private readonly IDatabase _database;
+        private readonly ICandlesCacheInitalizationService _cacheInitiaInitalizationService;
+        private readonly MarketType _market;
+        private readonly int _amountOfCandlesToStore;
+
+        public RedisCacheCaretaker(
+            ICandlesHistoryRepository historyRepository,
+            IDatabase database,
+            ICandlesCacheInitalizationService cacheInitiaInitalizationService,
+            MarketType market,
+            TimeSpan cacheCheckupPeriod,
+            int amountOfCandlesToStore,
+            ILog log)
+            : base(nameof(RedisCacheCaretaker), (int)cacheCheckupPeriod.TotalMilliseconds, log)
+        {
+            _historyRepository = historyRepository ?? throw new ArgumentNullException(nameof(historyRepository));
+            _database = database ?? throw new ArgumentNullException(nameof(database));
+            _cacheInitiaInitalizationService = cacheInitiaInitalizationService ?? throw new ArgumentNullException(nameof(cacheInitiaInitalizationService));
+            _market = market;
+            _amountOfCandlesToStore = amountOfCandlesToStore;
+        }
+
+        public override Task Execute()
+        {
+            // TODO: such an approach is Ok for the case of the single running service instance. But once we get
+            // TODO: a necessity to run more instances, the code below will provoke a problem.
+
+            TruncateCache();
+
+            ReloadCacheIfNeeded();
+            
+            return Task.CompletedTask;
+        }
+
+        private void TruncateCache()
+        {
+            // Shall not truncate cache while reloading data.
+            if (_cacheInitiaInitalizationService.InitializationState != CacheInitializationState.Idle)
+                return;
+
+            foreach (var assetId in _historyRepository.GetStoredAssetPairs())
+            {
+                foreach (var priceType in Constants.StoredPriceTypes)
+                {
+                    foreach (var timeInterval in Constants.StoredIntervals)
+                    {
+                        var key = RedisCandlesCacheService.GetKey(_market, assetId, priceType, timeInterval);
+
+                        _database.SortedSetRemoveRangeByRank(key, 0, -_amountOfCandlesToStore - 1, CommandFlags.FireAndForget);
+                    }
+                }
+            }
+        }
+
+        private void ReloadCacheIfNeeded()
+        {
+            var vkey = RedisCandlesCacheService.GetValidationKey(_market);
+            if (_database.KeyExists(vkey))
+                return;
+
+            _cacheInitiaInitalizationService.InitializeCacheAsync()
+                .GetAwaiter()
```

Comment:
Why containing method is not async?

Reply:
As we avoid starting new initialization until the previous ends, this does not make a principal difference. Anyway, timer handler execution runs in a separate thread and does not block enything but itself, afaik.
I may be wrong, of course. Please, tell me, if I am.

Reply:
Use the rule - if the method, in which async method is called, could be made async, make it async